[PLAT-50] GenericFileStore should validate URL

adborbas opened the pull request:
[Generic File Storage fails on PR builds with some env vars](https://bitrise.atlassian.net/browse/PLAT-50)

**Changes**

- Refactor `main.go` to extract functionality and make it testable. [refactor and adding tests](https://github.com/bitrise-steplib/steps-generic-file-storage/pull/7/commits/2ce947c735425a48d1d9c15d082a3b34ee5cd553)
- File downloading is done with [FileDownlader](https://github.com/bitrise-steplib/bitrise-step-export-universal-apk/blob/master/filedownloader/filedownloader.go)
- Update bitrise.yml to check output dir
- Extended `genericfilestorage.Storage`to validate URL's host and path before downloading file.


**Files changed**

- Everything except `/vendor` contents
- Ignore `Add * dep` commits

Commit:
add filedownloader dep

Commit:
Add mock dep

Commit:
Add dep require

Commit:
refactor and adding tests

Commit:
Adapt mockFileDownloader.go to new mock structure

Commit:
Update bitrise.yml to check output dir

Commit:
Validate URL host and path

adborbas committed:
Merge branch 'master' into plat_50_gfs

zsolt-marta-bitrise commented on genericfilestorage/fileprovider.go:
```diff
@@ -0,0 +1,52 @@
+)
+
+// OsEnvFileProvider ...
+type OsEnvFileProvider struct {
```
Given that `FileProvider` is defined elsewhere, shouldn't we call this file according to this, i.e. `osenvfileprovider`?

zsolt-marta-bitrise commented on genericfilestorage/storage.go:
```diff
@@ -0,0 +1,105 @@
+	}
+
+	fmt.Println()
+	log.Infof("Downloading %d files:", len(files))
```
Should we return early if `len(files) == 0`?

adborbas replied:
We could but it's still clear from the logs that there were 0 files to download. Plus there is not much performance cost of executing the next few lines. I would keep it as is, as this would rarely be the case that the step is added but there is no file to download.

zsolt-marta-bitrise replied:
Yeah, it's just that it looks a bit weird to say `"Downloading 0 files:"`. But I'm ok with leaving it here

zsolt-marta-bitrise commented on genericfilestorage/storage.go:
```diff
@@ -0,0 +1,105 @@
+	storage.logDebugPretty(fs)
+
+	if len(fs) > 0 {
+		log.Printf("  %s", files(fs).String(storage.envKey))
```
Do we need this along with `logDebugPretty `?

adborbas replied:
I'm not sure, to be honest, I've just copied this from the original implementation. The only difference I see between the two is `logDebugPretty` logs only if debug is enabled. Would you keep both with this in mind?

zsolt-marta-bitrise replied:
If it was like that before I'm ok with not changing functionality

Commit:
Fix BulkFileDownloader second call fails test

Commit:
Merge branch 'plat_50_gfs' of github.com:bitrise-steplib/steps-generic-file-storage into plat_50_gfs

zsolt-marta-bitrise commented on genericfilestorage/storage_test.go:
```diff
@@ -0,0 +1,104 @@
+	// Then
+	assert.NoError(t, err)
+	assert.NotEmpty(t, path)
+	assert.DirExists(t, path)
```
So the test will create files there? Should we clean them up?

adborbas replied:
nice catch

zsolt-marta-bitrise commented on genericfilestorage/storage_test.go:
```diff
@@ -0,0 +1,104 @@
+
+func givenFiles() []File {
+	return []File{
+		File{name: "name", url: "http://host.com/name"},
```
For more coverage I'd add more files here too

adborbas replied:
How would you increase coverage? There are different file URLs covered in `osenvfileprovider_test.go`.

zsolt-marta-bitrise replied:
Hm, if `GetFiles()` is tested with multiple files, I guess it doesn't help much

zsolt-marta-bitrise commented on main.go:
```diff
@@ -1,144 +1,65 @@
+	storageDir, err := storage.DownloadFiles()
 	if err != nil {
-		failf("Failed to create storage temp dir, error: %s", err)
+		failf("%s", "az")
```
I think you left this here :)

adborbas replied:
Yes, I did. 🤦‍♂️

zsolt-marta-bitrise reviewed (changes requested):
Nice job 👍 Thanks for improving testability 🏅 
Couple of additions/questions

Commit:
Review findings

- Proper error logging in main.go
- Assert errors equality
- rename fileprovider.go to isenvfileprovider.go
- clean up temp folder at end ot test

adborbas commented:
@zsolt-marta-bitrise I have addressed the findings. Could you please check again? Thanks

adborbas commented:
I have found an issue in bitirse.yml 😱. Will fix it soon.
UPDATE: Fixed now.

Commit:
Fix output checkin bitrise.yml
